draft-ietf-webtrans-http3.md

```diff
@@ -293,11 +293,19 @@ SHALL include a single choice from the client's list in that field. Servers MAY
 reject the request if the client did not include a suitable protocol.
 
 Both `WT-Available-Protocols` and `WT-Protocol` are Structured Fields
-{{!RFC8941}}. `WT-Available-Protocols` is a List of Tokens, and `WT-Protocol` is
-a Token. The token in the `WT-Protocol` response header field MUST be one of
-the tokens listed in `WT-Available-Protocols` of the request.  The semantics of
-individual token values is determined by the WebTransport resource in question
-and are not registered in IANA's "ALPN Protocol IDs" registry.
+{{!RFC8941}}. `WT-Available-Protocols` is a List, with the only valid value type
+being either String or Token.  `WT-Protocol` is defined as an item with the only
+valid type for values being String or Token. Any value type other than String or
+Token MUST be ignored.  String or Token values can be used interchangeably;
+decoded values are used to identify protocols. No semantics are defined for
+parameters on either field; parameters MUST be ignored.
```

Comment:
Is this required for some reasons that I am not structured-headers enough to understand?

Reply:
Structured fields is a bit loose.  There are assumed norms, but you can't just say "this is an item" without defining anything.  [Section 2 of RFC 9651](https://datatracker.ietf.org/doc/html/rfc9651#section-2) has a lot of stuff in it.  I've boiled all that down as much as I think we can get away with here.

Reply:
I am trying to understand how this went from tokens to tokens and strings; our implementation currently only accepts tokens, for what it's worth.

Reply:
Maybe a stupid question, but how are values separated? By commas as for WebSocket?

Reply:
And another question, I have set "WT-Available-Protocols" but google/quiche complains:
```
E0406 12:18:14.509503   84630 quic_spdy_stream.cc:1870] Invalid character in header name WT-Available-Protocols
```
reason is that quiche only allows lower case, so should it be written in lower case?

Reply:
Values are comma-separated, as defined in RFC 9651.

The upper/lower thing might be a bug in Google's implementation.  @DavidSchinazi @vasilvv are better positioned to answer that than I.

Reply:
A few points:
* HTTP header names[^1] are case-insensitive
* HTTP/2 and HTTP/3 require header names to be transmitted in lower-case
* The HTTP style guide recommends using mixed case for them in specifications

Because of all this, the spec should still say "WT-Available-Protocols" but if you're inside QUICHE you should pass in "wt-available-protocols". General-purpose embedders of quiche (like Chrome and Envoy) and have a conversion layer where they lowercase the header names before passing them in the quiche

[^1]: on the wind, I hear Mark whispering _fields! they're fields!_

Reply:
Thanks! That clarifies it, and I hope to finish implementing this feature in the library next weekend.

Reply:
We've already implemented parsing, but it only accepts tokens (not strings), which is why I am trying to understand why we're changing this.

Reply:
@martinthomson so can we change this text to say Tokens and be done with it?

Reply:
I should have said, RFC 9651 says:

> Note that Tokens are defined largely for compatibility with the data model of existing HTTP fields and may require additional steps to use in some implementations. As a result, new fields are encouraged to use Strings.

Aside from protecting Victor's implementation (which can change, last I checked), is there a reason not to accept strings?  I say that because there are a range of valid protocol identifiers that can't be carried in tokens.  Those include anything with uppercase letters (strangely, the registry has none of those; even HTTP/1.1 is identified in lowercase).

I understand why we wouldn't be able to *remove* tokens, but it seems like there's no reason to prohibit strings, where strings are what it should have been originally.